src/ccl_power.c

```diff
@@ -1037,130 +976,100 @@ static void ccl_cosmology_compute_power_bbks(ccl_cosmology * cosmo, int * status
-      cosmo->computed_power=true;
-    } else {
-      gsl_spline2d_free(log_power_lin_unnorm);
-      *status = CCL_ERROR_SPLINE;
-      ccl_cosmology_set_status_message(cosmo,"ccl_power.c: ccl_cosmology_compute_power_bbks(): Error creating log_power_lin spline\n");
-    }
-  }
 
-  if(*status){
-    free(x); free(y); free(a); free(y2d);
-    return;
+  if(*status==0) {
+    cosmo->data.p_lin=ccl_p2d_t_new(na,z,nk,x,y2d,1,2,ccl_p2d_cclgrowth,1,NULL,0,ccl_p2d_3,status);
+    cosmo->computed_power=true;
+    sigma8 = ccl_sigma8(cosmo,status);
+    cosmo->computed_power=false;
   }
 
-  // Calculate sigma8 for the unnormalized P(k), using the standard
-  // ccl_sigma8() function
-  double sigma8 = ccl_sigma8(cosmo,status);
-  cosmo->computed_power=false;
-
-  // Check that ccl_sigma8 didn't fail
-  if (!*status) {
-    
+  if(*status==0) {
     // Calculate normalization factor using computed value of sigma8, then
     // recompute P(k, a) using this normalization
-    double log_normalization_factor = 2*(log(cosmo->params.sigma8) - log(sigma8));
-    for (int i=0; i<nk; i++) {
-      y[i] += log_normalization_factor;
-    }
-    for (int j = 0; j < na; j++) {
-      double gfac = ccl_growth_factor(cosmo,a[j], status);
-      double g2 = 2.*log(gfac);
-      for (int i=0; i<nk; i++) {
-	y2d[j*nk+i] = y[i]+g2;
-      }
-    }
+    log_sigma8 = 2*(log(cosmo->params.sigma8) - log(sigma8));
+    for(int j=0;j<na*nk;j++)
+      y2d[j] += log_sigma8;
   }
-
-  //Check growth didn't fail
-  if (!*status) {
-    
-    gsl_spline2d * log_power_lin = gsl_spline2d_alloc(PLIN_SPLINE_TYPE, nk,na);
-    splinstatus = gsl_spline2d_init(log_power_lin, x, a, y2d,nk,na);
-
-    if (!splinstatus) {
-      cosmo->data.p_lin=log_power_lin;
-    } else {
-      gsl_spline2d_free(log_power_lin);
-      *status = CCL_ERROR_SPLINE;
-      ccl_cosmology_set_status_message(cosmo,"ccl_power.c: ccl_cosmology_compute_power_bbks(): Error creating log_power_lin spline\n");
-    }
+  
+  if(*status==0) {
+    // Free the previous P(k,a) spline, and allocate a new one to store the
+    // properly-normalized P(k,a)
+    ccl_p2d_t_free(cosmo->data.p_lin);
+    cosmo->data.p_lin=ccl_p2d_t_new(na,z,nk,x,y2d,1,2,ccl_p2d_cclgrowth,1,NULL,0,ccl_p2d_3,status);
   }
-
-  if(!*status){
-    // Allocate a 2D spline for the nonlinear P(k)
-    //[which is just a copy of the linear one for BBKS]
-    gsl_spline2d * log_power_nl = gsl_spline2d_alloc(PNL_SPLINE_TYPE, nk,na);
-    splinstatus = gsl_spline2d_init(log_power_nl, x, a, y2d,nk,na);
-    if (!splinstatus) {
-      cosmo->data.p_nl = log_power_nl;
-    } else {
-      gsl_spline2d_free(log_power_nl);
-      *status = CCL_ERROR_SPLINE;
-      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_bbks(): Error creating log_power_nl spline\n");
-    }
+  
+  if(*status==0) {
+    if(cosmo->config.baryons_power_spectrum_method==ccl_bcm)
+      correct_bcm(cosmo,na,z,nk,x,y2d,status);
```

Comment:
Again we should not be applying the BCM to the linear P(k).

Reply:
Agreed :-)

Reply:
Wow ok. So actually if you specify EH98 for the transfer function, you get that for the non-linear and linear power?

Reply:
@damonge pinging you to make sure you see this.

Reply:
discussed below.

Reply:
Sorry, didn't see this comment:

>So actually if you specify EH98 for the transfer function, you get that for the non-linear and linear power?

Yes, that's correct